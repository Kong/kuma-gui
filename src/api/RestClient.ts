import { makeRequest } from './makeRequest'

const DEFAULT_BASE_URL = import.meta.env.PROD ? window.location.origin : import.meta.env.VITE_KUMA_API_SERVER_URL

export class RestClient {
  /**
   * The API base URL. **Will always be stored without a trailing slash**.
   */
  _baseUrl: string = DEFAULT_BASE_URL

  /**
   * The absolute API base URL used in all requests. Includes its base path segment if one is set.
   */
  get baseUrl() {
    return this._baseUrl
  }

  /**
   * @param baseUrlOrPath the API base URL or the API base path
   */
  set baseUrl(baseUrlOrPath: string) {
    if (baseUrlOrPath.startsWith('http')) {
      this._baseUrl = trimTrailingSlashes(baseUrlOrPath)
    } else {
      const basePath = trimSlashes(baseUrlOrPath)
      this._baseUrl = [DEFAULT_BASE_URL, basePath].filter((segment) => segment !== '').join('/')
    }
  }

  /**
   * Performs a network request using the GET method.
   *
   * @returns the response’s de-serialized data.
   */
  async get(path: string, options?: RequestInit & { params?: any }): Promise<any> {
    const normalizedOptions = normalizeParameters(options)
    normalizedOptions.method = 'GET'

    const { data } = await this.raw(path, normalizedOptions)

    return data
  }

  /**
   * Performs a network request.
   *
   * @returns the response’s de-serialized data (when applicable) and the raw `Response` object.
   */
  async raw(urlOrPath: string, options?: RequestInit & { params?: any }): Promise<{ response: Response, data: any }> {
    const normalizedOptions = normalizeParameters(options)
    const url = urlOrPath.startsWith('http') ? urlOrPath : [this.baseUrl, urlOrPath].join('/')

    return makeRequest(url, normalizedOptions)
  }
}

function trimTrailingSlashes(str: string): string {
  return str.replace(/\/+$/, '')
}

function trimLeadingSlashes(str: string): string {
  return str.replace(/^\/+/, '')
}

function trimSlashes(str: string): string {
  return trimTrailingSlashes(trimLeadingSlashes(str))
}

function normalizeParameters(options?: RequestInit & { params?: any }): RequestInit & { params?: any } {
  const normalizedOptions: RequestInit & { params?: any } = options ?? {}

  // Turns parameter records (e.g. `{ tag: ['a', 'b'] }`) into parameter entries (e.g. `[['tag', 'a'], ['tag', 'b']]`) so they can be used to construct an appropriate `URLSearchParams` object. Without this normalization, the record form would produce `'tag=a%2Cb'` (i.e. comma-separated list of values) when we want `'tag=a&tag=b'` (i.e. one query parameter per value).
  if (normalizedOptions.params && !Array.isArray(normalizedOptions.params)) {
    const params = []

    for (const [param, value] of Object.entries(normalizedOptions.params)) {
      if (Array.isArray(value)) {
        for (const singleValue of value) {
          params.push([param, singleValue])
        }
      } else {
        params.push([param, value])
      }
    }

    normalizedOptions.params = params
  }

  return normalizedOptions
}
