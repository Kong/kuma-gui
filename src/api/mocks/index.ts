import { faker } from '@faker-js/faker'
import { rest, RestRequest } from 'msw'

import type Env from '@/services/env/Env'
import FakeKuma from '@/services/kuma-api/FakeKuma'
import type { Alias } from '@/services/utils'

export { fs } from './fs'

type Pager = (total: string | number, req: RestRequest, self: string) => {
  next: string | null,
  pageTotal: number,
  total: number,
  offset: number,
  size: number
}
const pager: Pager = (_total: string | number, req: RestRequest, self) => {
  const baseUrl = 'http://localhost:5681'
  const total = parseInt(`${_total}`)
  const query = req.url.searchParams
  const size = parseInt(query.get('size') || '10')
  const offset = parseInt(query.get('offset') || '1')
  const remaining = total - offset
  const pageTotal = Math.min(size, remaining)
  const next = remaining <= size ? null : `${baseUrl}${self}?offset=${offset + size}`
  return {
    next,
    pageTotal,
    total,
    offset,
    size,
  }
}

export type AEnv = Alias<Env['var']>
export type MockEnvKeys = keyof {
  FAKE_SEED: string
  KUMA_DATAPLANE_COUNT: string
}
export type AppEnvKeys = Parameters<AEnv>[0]
export type EndpointDependencies = {
  fake: FakeKuma
  pager: Pager,
  env: (key: AppEnvKeys | MockEnvKeys, d: string) => string
}
export type MockResponse = {
  headers: Record<string, string>
  body: string | Record<string, unknown>
}
export type MockResponder = (req: RestRequest) => MockResponse
export type FakeEndpoint = (deps: EndpointDependencies) => MockResponder

export function escapeRoute(route: string): string {
  return route.replaceAll('+', '\\+')
}
export const dependencies = {
  fake: new FakeKuma(faker),
  pager,
  env: (_key: AppEnvKeys | MockEnvKeys, d = '') => d,
}
export const fakeApi = (env: AEnv, fs: Record<string, FakeEndpoint>) => {
  const baseUrl = env('KUMA_API_URL')
  const mockEnv: EndpointDependencies['env'] = (key, d = '') => env(key as AppEnvKeys, d)

  return Object.entries(fs).map(([route, endpoint]) => {
    return rest.all(`${baseUrl}${escapeRoute(route)}`, async (req, res, ctx) => {
      const fetch = endpoint({
        ...dependencies,
        env: mockEnv,
      })
      const response = fetch(req)
      return res(
        ctx.status(parseInt(response.headers['Status-Code'] ?? '200')),
        ctx.json(response.body),
      )
    })
  })
}
export type FS = Record<string, FakeEndpoint>
