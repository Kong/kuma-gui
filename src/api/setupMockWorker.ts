import { setupWorker, MockedRequest, RestHandler } from 'msw'

/**
 * Sets up the msw-based mock server.
 */
export function setupMockWorker(handlers: RestHandler[]): void {
  const worker = setupWorker(...handlers)

  console.warn(
    '%c âœ¨You are mocking api requests.',
    'background: gray; color: white; display: block; padding: 0.25rem;',
  )

  worker.start({
    quiet: true,
    onUnhandledRequest(req: MockedRequest) {
      // Ignores warnings about unhandled requests.
      if (
        req.url.pathname.startsWith('/node_modules') ||
        req.url.pathname.startsWith('/src/assets') ||
        req.url.href.match(/\.(vue|ts|js|json)(\?.*)?$/)
      ) {
        return
      }

      console.warn('Found an unhandled %s request to %s', req.method, req.url.href)
    },
  })
}
