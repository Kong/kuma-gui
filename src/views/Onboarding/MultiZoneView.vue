<template>
  <OnboardingPage>
    <template #header>
      <OnboardingHeading title="Add zones" />
    </template>
    <template #content>
      <p class="text-center mb-4">
        A zone requires both the zone control plane and zone ingress. On Kubernetes, you run a single command to create both resources. On Universal, you must create them separately.
      </p>

      <KCard
        title="See the documentation for options to install:"
        border-variant="noBorder"
      >
        <template #body>
          <a
            target="_blank"
            class="external-link-code-block"
            :href="documentationLink"
          >
            {{ documentationLink }}
          </a>
        </template>
      </KCard>

      <div>
        <p class="text-center my-4">
          Zone status:
          <span
            v-if="hasZones"
            class="text-green-500"
            data-testid="zone-connected"
          >Connected</span>
          <span
            v-else
            class="text-red-500"
            data-testid="zone-disconnected"
          >Disconnected</span>
        </p>
        <p class="text-center mt-4 mb-10">
          Zone ingress status:
          <span
            v-if="hasZoneIngresses"
            class="text-green-500"
            data-testid="zone-ingress-connected"
          >Connected</span>
          <span
            v-else
            class="text-red-500"
            data-testid="zone-ingress-disconnected"
          >Disconnected</span>
        </p>
        <div
          v-if="!hasZoneIngresses || !hasZones"
          class="flex justify-center"
        >
          <LoadingBox />
        </div>
      </div>
    </template>

    <template #navigation>
      <OnboardingNavigation
        next-step="onboarding-create-mesh"
        previous-step="onboarding-configuration-types"
        :should-allow-next="servicesOnline"
      />
    </template>
  </OnboardingPage>
</template>

<script>
import { mapGetters } from 'vuex'

import Kuma from '@/services/kuma'
import LoadingBox from '@/components/LoadingBox.vue'
import OnboardingNavigation from '@/views/Onboarding/components/OnboardingNavigation.vue'
import OnboardingHeading from '@/views/Onboarding/components/OnboardingHeading.vue'
import OnboardingPage from '@/views/Onboarding/components/OnboardingPage.vue'

const LONG_POOLING_INTERVAL = 1000

export default {
  name: 'MultiZoneView',
  components: {
    OnboardingNavigation,
    OnboardingHeading,
    OnboardingPage,
    LoadingBox,
  },
  metaInfo() {
    return {
      title: 'Multizone',
    }
  },
  data() {
    return {
      hasZones: false,
      hasZoneIngresses: false,
      zoneTimeout: null,
      zoneIngressTimeout: null,
    }
  },
  computed: {
    servicesOnline() {
      return this.hasZoneIngresses && this.hasZones
    },
    ...mapGetters({
      kumaDocsVersion: 'config/getKumaDocsVersion',
    }),
    documentationLink() {
      return `https://kuma.io/docs/${this.kumaDocsVersion}/deployments/multi-zone/#zone-control-plane`
    },
  },

  created() {
    this.getZoneIngresses()
    this.getZones()
  },
  destroyed() {
    clearTimeout(this.zoneTimeout)
    clearTimeout(this.zoneIngressTimeout)
  },
  methods: {
    async getZones() {
      try {
        const { total } = await Kuma.getZones()

        this.hasZones = total > 0
      } catch (e) {
        console.error(e)
      }

      if (!this.hasZones) {
        this.zoneTimeout = setTimeout(() => this.getZones(), LONG_POOLING_INTERVAL)
      }
    },
    async getZoneIngresses() {
      try {
        const { total } = await Kuma.getAllZoneIngressOverviews()

        this.hasZoneIngresses = total > 0
      } catch (e) {
        console.error(e)
      }

      if (!this.hasZoneIngresses) {
        this.zoneIngressTimeout = setTimeout(() => this.getZoneIngresses(), LONG_POOLING_INTERVAL)
      }
    },
  },
}
</script>
