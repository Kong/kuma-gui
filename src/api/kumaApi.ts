import { RestClient } from './RestClient'
import {
  ApiKindListResponse,
  ApiListResponse,
  DataPlaneOverviewParameters,
  PaginatedApiListResponse,
  PaginationParameters,
} from '@/types/api.d'
import {
  DataPlane,
  DataPlaneOverview,
  DataplaneRule,
  ExternalService,
  GlobalInsights,
  Info,
  Mesh,
  MeshGatewayDataplane,
  MeshInsight,
  PolicyEntity,
  PolicyType,
  ServiceInsight,
  SidecarDataplane,
  Zone,
  ZoneOverview,
} from '@/types/index.d'
import { ClientConfigInterface } from '@/store/modules/config/config.types'

class KumaApi {
  client: RestClient

  constructor() {
    this.client = new RestClient()
  }

  get baseUrl() {
    return this.client.baseUrl
  }

  /**
   * Sets the API base URL for all network requests.
   *
   * URLs for requests will be constructed in the form `${baseUrl}/${path}`.
   */
  setBaseUrl(baseUrl: string): void {
    this.client.baseUrl = baseUrl
  }

  getInfo(): Promise<Info> {
    return this.client.get('')
  }

  async getStatus(): Promise<'OK' | null> {
    try {
      const { response } = await this.client.raw('')

      return response.status === 200 ? 'OK' : null
    } catch {
      return null
    }
  }

  async getLatestVersion(): Promise<string> {
    return this.client.get(import.meta.env.VITE_VERSION_URL)
  }

  getConfig(): Promise<ClientConfigInterface> {
    return this.client.get('config')
  }

  /**
   * Retrieves a list of known policy definitions.
   */
  getPolicyTypes(): Promise<{ policies: PolicyType[] }> {
    return this.client.get('policies')
  }

  getGlobalInsights(): Promise<GlobalInsights> {
    return this.client.get('global-insights')
  }

  getZones(params?: PaginationParameters): Promise<PaginatedApiListResponse<Zone>> {
    return this.client.get('zones', { params })
  }

  getZone({ name }: { name: string }, params?: any): Promise<Zone> {
    return this.client.get(`zones/${name}`, { params })
  }

  getAllZoneOverviews(params?: PaginationParameters): Promise<PaginatedApiListResponse<ZoneOverview>> {
    return this.client.get('zones+insights', { params })
  }

  getZoneOverview({ name }: { name: string }, params?: PaginationParameters): Promise<ZoneOverview> {
    return this.client.get(`zones+insights/${name}`, { params })
  }

  /**
   * Fetches additional data like xDS configuration, envoy stats, or envoy clusters.
   */
  getZoneIngressData({ zoneIngressName, dataPath }: { zoneIngressName: string, dataPath: 'xds' | 'stats' | 'clusters' }, params?: any): Promise<string> {
    return this.client.get(`zoneingresses/${zoneIngressName}/${dataPath}`, { params })
  }

  getAllZoneIngressOverviews(params?: PaginationParameters): Promise<any> {
    return this.client.get('zoneingresses+insights', { params })
  }

  getZoneIngressOverview({ name }: { name: string }, params?: PaginationParameters): Promise<any> {
    return this.client.get(`zoneingresses+insights/${name}`, { params })
  }

  /**
   * Fetches additional data like xDS configuration, envoy stats, or envoy clusters.
   */
  getZoneEgressData({ zoneEgressName, dataPath }: { zoneEgressName: string, dataPath: 'xds' | 'stats' | 'clusters' }, params?: any): Promise<string> {
    return this.client.get(`zoneegresses/${zoneEgressName}/${dataPath}`, { params })
  }

  getAllZoneEgressOverviews(params?: PaginationParameters): Promise<any> {
    return this.client.get('zoneegressoverviews', { params })
  }

  getZoneEgressOverview({ name }: { name: string }, params?: PaginationParameters): Promise<any> {
    return this.client.get(`zoneegressoverviews/${name}`, { params })
  }

  getAllMeshes(params?: PaginationParameters): Promise<PaginatedApiListResponse<Mesh>> {
    return this.client.get('meshes', { params })
  }

  getMesh({ name }: { name: string }, params?: any): Promise<Mesh> {
    return this.client.get(`meshes/${name}`, { params })
  }

  getAllMeshInsights(params?: PaginationParameters): Promise<PaginatedApiListResponse<MeshInsight>> {
    return this.client.get('mesh-insights', { params })
  }

  getMeshInsights({ name }: { name: string }, params?: any): Promise<MeshInsight> {
    return this.client.get(`mesh-insights/${name}`, { params })
  }

  getAllDataplanes(params?: PaginationParameters): Promise<PaginatedApiListResponse<DataPlane>> {
    return this.client.get('dataplanes', { params })
  }

  getDataplaneFromMesh({ mesh, name }: { mesh: string, name: string }, params?: any): Promise<DataPlane> {
    return this.client.get(`meshes/${mesh}/dataplanes/${name}`, { params })
  }

  getAllDataplaneOverviews(params?: DataPlaneOverviewParameters): Promise<PaginatedApiListResponse<DataPlaneOverview>> {
    return this.client.get('dataplanes+insights', { params })
  }

  getAllDataplaneOverviewsFromMesh({ mesh }: { mesh: string }, params?: DataPlaneOverviewParameters): Promise<PaginatedApiListResponse<DataPlaneOverview>> {
    return this.client.get(`meshes/${mesh}/dataplanes+insights`, { params })
  }

  getDataplaneOverviewFromMesh({ mesh, name }: { mesh: string, name: string }, params?: any): Promise<DataPlaneOverview> {
    return this.client.get(`meshes/${mesh}/dataplanes+insights/${name}`, { params })
  }

  getSidecarDataplanePolicies({ mesh, name }: { mesh: string; name: string }, params?: any): Promise<ApiKindListResponse<SidecarDataplane>> {
    return this.client.get(`meshes/${mesh}/dataplanes/${name}/policies`, { params })
  }

  getMeshGatewayDataplane({ mesh, name }: { mesh: string; name: string }, params?: any): Promise<MeshGatewayDataplane> {
    return this.client.get(`meshes/${mesh}/dataplanes/${name}/policies`, { params })
  }

  getDataplaneRules({ mesh, name }: { mesh: string; name: string }, params?: any): Promise<ApiListResponse<DataplaneRule>> {
    return this.client.get(`meshes/${mesh}/dataplanes/${name}/rules`, { params })
  }

  /**
   * Fetches additional data like xDS configuration, envoy stats, or envoy clusters.
   */
  getDataplaneData({ mesh, dppName, dataPath }: { mesh: string, dppName: string, dataPath: 'xds' | 'stats' | 'clusters' }, params?: any): Promise<string> {
    return this.client.get(`meshes/${mesh}/dataplanes/${dppName}/${dataPath}`, { params })
  }

  getAllServiceInsights(params?: PaginationParameters): Promise<PaginatedApiListResponse<ServiceInsight>> {
    return this.client.get('service-insights', { params })
  }

  getAllServiceInsightsFromMesh({ mesh }: { mesh: string }, params?: PaginationParameters): Promise<PaginatedApiListResponse<ServiceInsight>> {
    return this.client.get(`meshes/${mesh}/service-insights`, { params })
  }

  getServiceInsight({ mesh, name }: { mesh: string, name: string }, params?: any): Promise<ServiceInsight> {
    return this.client.get(`meshes/${mesh}/service-insights/${name}`, { params })
  }

  getAllExternalServices(params?: PaginationParameters): Promise<PaginatedApiListResponse<ExternalService>> {
    return this.client.get('external-services', { params })
  }

  getAllExternalServicesFromMesh({ mesh }: { mesh: string }, params?: PaginationParameters): Promise<PaginatedApiListResponse<ExternalService>> {
    return this.client.get(`meshes/${mesh}/external-services`, { params })
  }

  getExternalService({ mesh, name }: { mesh: string, name: string }, params?: any): Promise<ExternalService> {
    return this.client.get(`meshes/${mesh}/external-services/${name}`, { params })
  }

  getPolicyConnections({ mesh, policyType, policyName }: { mesh: string; policyType: string; policyName: string }, params?: PaginationParameters): Promise<PaginatedApiListResponse<any>> {
    return this.client.get(`meshes/${mesh}/${policyType}/${policyName}/dataplanes`, { params })
  }

  getAllPolicyEntities({ path }: { path: string }, params?: PaginationParameters): Promise<PaginatedApiListResponse<PolicyEntity>> {
    return this.client.get(path, { params })
  }

  getAllPolicyEntitiesFromMesh({ mesh, path }: { mesh: string, path: string }, params?: PaginationParameters): Promise<PaginatedApiListResponse<PolicyEntity>> {
    return this.client.get(`meshes/${mesh}/${path}`, { params })
  }

  getSinglePolicyEntity({ mesh, path, name }: { mesh: string, path: string, name: string }, params?: any): Promise<PolicyEntity> {
    return this.client.get(`meshes/${mesh}/${path}/${name}`, { params })
  }
}

export const kumaApi = new KumaApi()
