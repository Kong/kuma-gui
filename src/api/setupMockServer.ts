import { setupServer, SetupServer } from 'msw/node'

import { setupHandlers } from './mocks'

/**
 * Sets up a mock server for the use in tests.
 *
 * **IMPORTANT**: Do not import this file in the regular application. Since it imports `msw/node`, this will cause the application to break because it will try to import (require, actually) Node built-ins which arenâ€™t available in browser environments.
 */
export function setupMockServer(url: string = ''): SetupServer {
  return setupServer(...setupHandlers(url))
}
