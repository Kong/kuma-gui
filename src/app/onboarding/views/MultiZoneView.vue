<template>
  <OnboardingPage>
    <template #header>
      <OnboardingHeading>
        <template #title>
          Add zones
        </template>
      </OnboardingHeading>
    </template>

    <template #content>
      <p class="mb-4 text-center">
        A zone requires both the zone control plane and zone ingress. On Kubernetes, you run a single command to create both resources. On Universal, you must create them separately.
      </p>

      <p class="mb-4 text-center">
        <b>See <a
          :href="`${env('KUMA_DOCS_URL')}/deployments/multi-zone/?${env('KUMA_UTM_QUERY_PARAMS')}#zone-control-plane`"
          target="_blank"
        >the documentation for options to install</a>.</b>
      </p>

      <div>
        <p class="status-box mt-4">
          Zone status:

          <span
            v-if="hasZones"
            class="status--is-connected"
            data-testid="zone-connected"
          >Connected</span>

          <span
            v-else
            class="status--is-disconnected"
            data-testid="zone-disconnected"
          >Disconnected</span>
        </p>

        <p class="status-box mt-4">
          Zone ingress status:

          <span
            v-if="hasZoneIngresses"
            class="status--is-connected"
            data-testid="zone-ingress-connected"
          >Connected</span>

          <span
            v-else
            class="status--is-disconnected"
            data-testid="zone-ingress-disconnected"
          >Disconnected</span>
        </p>

        <div
          v-if="!hasZoneIngresses || !hasZones"
          class="status-loading-box mt-4"
        >
          <LoadingBox />
        </div>
      </div>
    </template>

    <template #navigation>
      <OnboardingNavigation
        next-step="onboarding-create-mesh"
        previous-step="onboarding-configuration-types"
        :should-allow-next="hasZones && hasZoneIngresses"
      />
    </template>
  </OnboardingPage>
</template>

<script lang="ts" setup>
import { onUnmounted, ref } from 'vue'

import { kumaApi } from '@/api/kumaApi'
import { useEnv } from '@/utilities'
import LoadingBox from '../components/LoadingBox.vue'
import OnboardingHeading from '../components/OnboardingHeading.vue'
import OnboardingNavigation from '../components/OnboardingNavigation.vue'
import OnboardingPage from '../components/OnboardingPage.vue'

const env = useEnv()

const LONG_POOLING_INTERVAL = 1000

const hasZones = ref(false)
const hasZoneIngresses = ref(false)
const zoneTimeout = ref<number | null>(null)
const zoneIngressTimeout = ref<number | null>(null)

onUnmounted(function () {
  clearZoneTimeout()
  clearZoneIngressTimeout()
})

getZones()
getZoneIngresses()

async function getZones() {
  try {
    const { total } = await kumaApi.getZones()

    hasZones.value = total > 0
  } catch (error) {
    console.error(error)
  } finally {
    if (!hasZones.value) {
      clearZoneTimeout()
      zoneTimeout.value = window.setTimeout(getZones, LONG_POOLING_INTERVAL)
    }
  }
}

async function getZoneIngresses() {
  try {
    const { total } = await kumaApi.getAllZoneIngressOverviews()

    hasZoneIngresses.value = total > 0
  } catch (error) {
    console.error(error)
  } finally {
    if (!hasZoneIngresses.value) {
      clearZoneIngressTimeout()
      zoneIngressTimeout.value = window.setTimeout(getZoneIngresses, LONG_POOLING_INTERVAL)
    }
  }
}

function clearZoneTimeout() {
  if (zoneTimeout.value !== null) {
    window.clearTimeout(zoneTimeout.value)
  }
}

function clearZoneIngressTimeout() {
  if (zoneIngressTimeout.value !== null) {
    window.clearTimeout(zoneIngressTimeout.value)
  }
}
</script>

<style lang="scss" scoped>
.status-box {
  text-align: center;
}

.status--is-connected {
  color: var(--green-500);
}

.status--is-disconnected {
  color: var(--red-500);
}

.status-loading-box {
  display: flex;
  justify-content: center;
}
</style>
