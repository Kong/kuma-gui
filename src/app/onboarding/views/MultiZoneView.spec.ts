import { describe, expect, jest, test } from '@jest/globals'
import { flushPromises, mount } from '@vue/test-utils'

import MultiZoneView from './MultiZoneView.vue'
import { kumaApi } from '@/api/kumaApi'

function renderComponent() {
  return mount(MultiZoneView)
}

describe('MultiZoneView.vue', () => {
  test('renders snapshot', () => {
    const wrapper = renderComponent()

    expect(wrapper.element).toMatchSnapshot()
  })

  test('detects resources on call and allow to proceed', async () => {
    const wrapper = renderComponent()

    await flushPromises()

    expect(wrapper.find('[data-testid="zone-connected"]').exists()).toBe(true)
    expect(wrapper.find('[data-testid="zone-ingress-connected"]').exists()).toBe(true)
    expect(wrapper.find('[data-testid="onboarding-next-button"]').exists()).toBe(true)
    expect(wrapper.find('[data-testid="loading"]').exists()).toBe(false)
  })

  test('refetch resources if any not available', async () => {
    jest.useFakeTimers()
    jest
      .spyOn(kumaApi, 'getAllZoneIngressOverviews')
      .mockResolvedValueOnce({
        total: 0,
        items: [],
        next: null,
      })
      .mockResolvedValueOnce({
        total: 1,
        items: [],
        next: '',
      })

    const wrapper = renderComponent()

    await flushPromises()

    expect(wrapper.find('[data-testid="zone-connected"]').exists()).toBe(true)
    expect(wrapper.find('[data-testid="zone-ingress-disconnected"]').exists()).toBe(true)
    expect(wrapper.find('[data-testid="loading"]').exists()).toBe(true)

    jest.runAllTimers()
    await flushPromises()

    expect(wrapper.find('[data-testid="zone-ingress-connected"]').exists()).toBe(true)
    expect(kumaApi.getAllZoneIngressOverviews).toHaveBeenCalledTimes(2)
    jest.useRealTimers()
  })
})
