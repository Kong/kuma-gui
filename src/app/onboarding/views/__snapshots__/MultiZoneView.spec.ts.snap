// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiZoneView.vue renders snapshot 1`] = `
<div>
  <div
    class="onboarding-container"
  >
    <div
      class="onboarding-container__header"
    >
      
      <div
        class="onboarding-heading"
      >
        <h1
          class="onboarding-title"
        >
          
           Add zones 
          
        </h1>
        <!--v-if-->
      </div>
      
    </div>
    <div
      class="onboarding-container__content"
    >
      <div
        class="onboarding-container__inner-content"
      >
        
        <p
          class="mb-4 text-center"
        >
           A zone requires both the zone control plane and zone ingress. On Kubernetes, you run a single command to create both resources. On Universal, you must create them separately. 
        </p>
        <p
          class="mb-4 text-center"
        >
          <b>
            See 
            <a
              href="https://kuma.io/docs/1.7.x/deployments/multi-zone/?utm_source=Kuma&utm_medium=Kuma#zone-control-plane"
              target="_blank"
            >
              the documentation for options to install
            </a>
            .
          </b>
        </p>
        <div>
          <p
            class="status-box mt-4"
          >
             Zone status: 
            <span
              class="status--is-disconnected"
              data-testid="zone-disconnected"
            >
              Disconnected
            </span>
          </p>
          <p
            class="status-box mt-4"
          >
             Zone ingress status: 
            <span
              class="status--is-disconnected"
              data-testid="zone-ingress-disconnected"
            >
              Disconnected
            </span>
          </p>
          <div
            class="status-loading-box mt-4"
          >
            <div
              class="loading"
              data-testid="loading"
            >
              <div />
              <div />
              <div />
              <div />
              <div />
            </div>
          </div>
        </div>
        
      </div>
    </div>
    <div
      class="mt-4"
    >
      
      <div
        class="onboarding-actions"
      >
        <a
          class="k-button medium rounded secondary"
          data-testid="onboarding-previous-button"
          href="/onboarding/configuration-types"
          type="button"
        >
          
          <!---->
          
          
           Back 
          
          <!---->
        </a>
        <div
          class="button-list"
        >
          <a
            class="k-button medium rounded outline"
            data-testid="onboarding-skip-button"
            href="/"
            type="button"
          >
            
            <!---->
            
            
             Skip setup 
            
            <!---->
          </a>
          <a
            class="k-button medium rounded primary"
            data-testid="onboarding-next-button"
            disabled="true"
            href="/onboarding/create-mesh"
            type="button"
          >
            
            <!---->
            
            
            Next
            
            <!---->
          </a>
        </div>
      </div>
      
    </div>
  </div>
  <div
    class="background-image"
  />
</div>
`;
